File: php/src/Transaction.php
<?php

declare(strict_types=1);

namespace Inillucent;

/**
 * One transaction, held open while the caller decides whether to commit.
 *
 * This is a handle rather than a pair of calls because a check on what a write
 * did has to happen before the commit. A postcondition tested afterwards is a
 * report about something that has already happened.
 *
 * Letting it go out of scope without committing rolls it back.
 */
final class Transaction
{
    /** @var int[] how many rows each statement in this transaction changed, in order */
    public array $affected = [];

    private mixed $handle;

    /**
     * @param mixed $handle the C transaction handle
     */
    public function __construct(mixed $handle)
    {
        $this->handle = $handle;
    }

    /**
     * Runs one statement inside the transaction and returns the rows it changed.
     *
     * A failure rolls the whole transaction back before it throws, so a caller
     * that stops at the first error has already undone everything.
     *
     * @param string $sql the statement to run
     */
    public function execute(string $sql): int
    {
        $ffi = Driver::ffi();
        $changed = $ffi->new('unsigned long long[1]');
        $error = $ffi->new('inillucent_error*[1]');
        InillucentException::check(
            $ffi->inillucent_txn_execute($this->handle, $sql, $changed, $error),
            $error
        );
        $this->affected[] = $changed[0];
        return $changed[0];
    }

    /** Commits the transaction. The handle is spent either way. */
    public function commit(): void
    {
        if ($this->handle === null) {
            return;
        }
        $ffi = Driver::ffi();
        $committing = $this->handle;
        $this->handle = null;
        $error = $ffi->new('inillucent_error*[1]');
        $status = $ffi->inillucent_txn_commit($committing, $error);
        try {
            InillucentException::check($status, $error);
        } finally {
            $ffi->inillucent_txn_rollback($committing);
        }
    }

    /** Rolls the transaction back and frees it. Rolling back twice is safe. */
    public function rollback(): void
    {
        if ($this->handle === null) {
            return;
        }
        Driver::ffi()->inillucent_txn_rollback($this->handle);
        $this->handle = null;
    }

    /** Rolls back when the transaction was never committed. */
    public function __destruct()
    {
        $this->rollback();
    }
}

File: java/src/main/java/com/inillucent/Transaction.java
package com.inillucent;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.ArrayList;
import java.util.List;

/**
 * One transaction, held open while the caller decides whether to commit.
 *
 * The caller holds it open, runs statements, reads how many rows each one
 * changed, and only then commits. A check made after the commit cannot stop the
 * write it was checking.
 *
 * It is AutoCloseable and closing rolls back, so a try with resources that ends
 * without a commit undoes its work, which is what an early return means.
 */
public final class Transaction implements AutoCloseable {

    private final Driver driver;
    private final List<Long> affected = new ArrayList<>();
    private MemorySegment handle;

    Transaction(Driver driver, MemorySegment handle) {
        this.driver = driver;
        this.handle = handle;
    }

    /**
     * Runs one statement inside the transaction and returns the rows it changed.
     *
     * A failure rolls the whole transaction back before it throws, so a caller
     * that stops at the first error has already undone everything.
     *
     * @param sql - the statement to run
     */
    public long execute(String sql) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment changed = arena.allocate(ValueLayout.JAVA_LONG);
            MemorySegment error = arena.allocate(ValueLayout.ADDRESS);
            int status = driver.callInt("inillucent_txn_execute",
                handle, arena.allocateFrom(sql), changed, error);
            Check.status(driver, status, error);
            long count = changed.get(ValueLayout.JAVA_LONG, 0);
            affected.add(count);
            return count;
        }
    }

    /** How many rows each statement in this transaction changed, in order. */
    public List<Long> affected() {
        return List.copyOf(affected);
    }

    /** Commits the transaction. The handle is spent either way. */
    public void commit() {
        if (handle == null) {
            return;
        }
        MemorySegment committing = handle;
        handle = null;
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment error = arena.allocate(ValueLayout.ADDRESS);
            int status = driver.callInt("inillucent_txn_commit", committing, error);
            try {
                Check.status(driver, status, error);
            } finally {
                driver.callVoid("inillucent_txn_rollback", committing);
            }
        }
    }

    /** Rolls the transaction back and frees it. Rolling back twice is safe. */
    public void rollback() {
        if (handle == null) {
            return;
        }
        driver.callVoid("inillucent_txn_rollback", handle);
        handle = null;
    }

    /** Rolls back when the transaction was not committed. */
    @Override
    public void close() {
        rollback();
    }
}
